feat: add rfe posts (#105)
mlr-org/mlr3website: mlr-org/gallery/optimization/2023-02-07-recursive-feature-elimination/index.qmd

---
title: "Recursive Feature Elimination on the Sonar Data Set"
description: |
  Utilize the built-in feature importance of models.
categories:
  - feature selection
  - classification
author:
  - name: Marc Becker

>     url: https://github.com/be-marc

Reviewer: ```suggestion
    url: https://github.com/be-marc
  - name: Sebastian Fischer   
    url: https://github.com/sebffischer
```

set.seed(1)
future::plan("multisession")
```

# Scope

Feature selection is the process of finding an optimal set of features to improve the performance, interpretability and robustness of machine learning algorithms.
In this article, we introduce the wrapper feature selection method *Recursive Feature Elimination*.
[Wrapper methods](https://mlr3book.mlr-org.com/feature-selection.html#fs-wrapper) iteratively select features that optimize a performance measure.
As an example, we will search for the optimal set of features for a `r ref("mlr_learners_classif.gbm", "gradient boosting machine")` and `r ref("mlr_learners_classif.svm", "support vector machine")` on the `r ref("mlr_tasks_pima", text = "Sonar")` data set.
We assume that you are already familiar with the basic building blocks of the [mlr3 ecosystem](https://mlr-org.com/ecosystem.html).
If you are new to feature selection, we recommend reading the [feature selection chapter](https://mlr3book.mlr-org.com/feature-selection.html) of the mlr3book first.

# Recursive Feature Elimination

Recursive Feature Elimination (RFE) is a widely used feature selection method for high-dimensional data sets.

> The idea is to iteratively remove the weakest feature from a model until the desired number of features is reached.

Reviewer: ```suggestion
The idea is to --starting with all features -- iteratively remove the weakest feature from a model until the desired number of features is reached.
```

For a full list of supported learners, see the learner page on the [mlr-org website](https://mlr-org.com/learners.html) and search for `"importance"`.

@guyon_gene_2002 developed the RFE algorithm for SVMs (SVM-RFE) to select informative genes in cancer classification.
The importance of the features is given by the weight vector of a linear support vector machine.
This method was later extended to other machine learning algorithms.
The only requirement is that the models can internally measure the feature importance.
The random forest algorithm offers multiple options for measuring feature importance.
The commonly used methods are the mean decrease in accuracy (MDA) and the mean decrease in impurity (MDI).
The MDA measures the decrease in accuracy for a feature if it was randomly permuted in the out-of-bag sample.
The MDI is the total reduction in node impurity when the feature is used for splitting.
Gradient boosting algorithms like `r ref("mlr_learners_classif.xgboost", "XGBoost")`, `r ref("mlr_learners_classif.lightgbm", "LightGBM")` and `r ref("mlr_learners_classif.gbm", "GBM")` use similar methods to measure the importance of the features.

> Resampling strategies can be combined with the algorithm in different ways.

Reviewer: ```suggestion
Resampling strategies can be combined with the algorithm in various ways.
```

> The frameworks scikit-learn [@pedregosa_scikit-learn_2011] and caret [@kuhn_building_2008] implement a variant called recursive feature elimination with cross-validation (RFE-CV) that estimates the optimal number of features with cross-validation first.

Reviewer: Dieser Satz ist ein bissl random. Ist mir nicht klar was man dadurch verstehen soll.

Reviewer: Ok kommt weiter unten, aber wenn mlr3 das auch implementiert dann sollte das hier schon stehen, ansonsten ist es weird dass gleich weiter unten gesagt wird, dass mlr3 das auch kann.

Then one more RFE is carried out on the complete dataset with the optimal number of features as the final feature set size.

> The RFE implementation in mlr3 can rank and aggregate importance scores across resampling iterations.

Reviewer: Der Satz macht nicht gan klar wie sich dass dann von RFE-CV unterscheidet. Würde noch einen Satz hinzuschreiben.

We will explore both variants in more detail below.

`r ref_pkg("mlr3fselect")` is the feature selection package of the [mlr3 ecosystem](https://mlr-org.com/ecosystem.html).
It implements the `r ref("mlr_fselectors_rfe", "RFE")` and `r ref("mlr_fselectors_rfecv", "RFE-CV")` algorithm.
We load all packages of the ecosystem with the `r ref("mlr3verse")` package.

```{r 2023-02-07-recursive-feature-elimination-002}
#| message: false

library(mlr3verse)
```

We retrieve the `r ref("mlr_fselectors_rfe", "RFE")` optimizer with the `r ref("fs()")` function.

```{r 2023-02-07-recursive-feature-elimination-003}
optimizer = fs("rfe",
  n_features = 1,
  feature_number = 1,
  aggregation = "rank")
```

The algorithm has multiple control parameters.

> The optimizer stops when the number of features equals `n_features`.

Reviewer: ```suggestion
The optimizer stops when the number of features equals `n_features` (remember that the starting point is the full feature set and we iteratively remove features).
```

> The parameters `feature_number`, `feature_fraction` and `subset_size` determine the number of features that are removed in each iteration.

Reviewer: ```suggestion
The mutually exclusive parameters `feature_number`, `feature_fraction` and `subset_size` control the number of features that are removed in each iteration: 
* `feature_number` option removes a fixed number of features in each iteration
* `feature_fraction` removes a fraction of the features
* `subset_size` argument is a vector that specifies exactly how many features are removed in each iteration.
```

> The `feature_number` option removes a fixed number of features in each iteration, whereas `feature_fraction` removes a fraction of the features.
> The `subset_size` argument is a vector that specifies exactly how many features are removed in each iteration.

Reviewer: ```suggestion
```

> The parameters are mutually exclusive and the default is `feature_fraction = 0.5`.

Reviewer: ```suggestion
The default is `feature_fraction = 0.5`.
```

> Usually, RFE fits a new model in each resampling iteration and calculates the feature importance again.

Reviewer: ```suggestion
Usually, RFE fits a new model in each resampling iteration and calculates the feature importance.
```

Reviewer: Was heißt hier usually?

> We can deactivate this behavior by setting `recursive = FALSE`.

Reviewer: was passiert dann?

When running an RFE with a resampling strategy like cross-validation, multiple models and importance scores are generated.
The `aggregation` parameter determines how the importance scores are aggregated.
The option `"rank"` ranks the importance scores in each iteration and then averages the ranks of the features.

> The feature with the lowest average rank is removed.

Reviewer: ```suggestion
The features with the lowest average rank is removed, where the number of removed features is determined by the control parameters mentioned above.
```

The option `"mean"` averages the importance scores of the features across the iterations.
The `"mean"` should only be used if the learner's importance scores can be reasonably averaged.

# Task

The objective of the `r ref("mlr_tasks_sonar", text = "Sonar")` data set is to predict whether a sonar signal bounced off a metal cylinder or a rock.
The data set includes 60 numerical features (see @fig-features).

```{r 2023-02-07-recursive-feature-elimination-004}
task = tsk("sonar")
```

> ```{r 2023-02-07-recursive-feature-elimination-005}

Reviewer: Würde hier eventuell nur ein paar exemplarisch zeigen.

  terminator = trm("none"),
  callback = clbk("mlr3fselect.svm_rfe"))

optimizer$optimize(instance)
```

@fig-svm-2 shows a similar optimization curve as @fig-svm but with fewer evaluated feature sets.

```{r 2023-02-07-recursive-feature-elimination-017}
#| code-fold: true
#| label: fig-svm-2
#| fig-cap: "Optimization path of the feature selection."

library(viridisLite)
library(mlr3misc)

data = as.data.table(instance$archive)
data[, n:= map_int(importance, length)]

ggplot(data, aes(x = n, y = classif.auc)) +
  geom_line(
    color = viridis(1, begin = 0.5),
    linewidth = 1) +
  geom_point(
    fill = viridis(1, begin = 0.5),
    shape = 21,
    size = 3,
    stroke = 0.5,
    alpha = 0.8) +
  xlab("Number of Features") +
  scale_x_reverse() +
  theme_minimal()
```

# Recursive Feature Elimination with Cross Validation

RFE-CV estimates the optimal number of features before selecting a feature set.
For this, an RFE is run in each resampling iteration and the number of features with the best mean performance is selected (see @fig-flowchart).
Then one more RFE is carried out on the complete dataset with the optimal number of features as the final feature set size.

```{mermaid}

> %%| label: fig-flowchart

Reviewer: MEGA NICE

The feature subsets of the final RFE run have no value in the `"iteration"` column because they were evaluated on the complete data set.

```{r 2023-02-07-recursive-feature-elimination-021}
#| column: body-outset

as.data.table(instance$archive)[, list(features, classif.auc, iteration, importance)]
```

# Final Model

The learner we use to make predictions on new data is called the final model.
The final model is trained with the optimal feature set on the full data set.
The optimal set consists of 19 features and is stored in `instance$result_feature_set`.
We subset the task to the optimal feature set and train the learner.

```{r 2023-02-07-recursive-feature-elimination-022}
#| output: false

task$select(instance$result_feature_set)
learner$train(task)
```

The trained model can now be used to predict new, external data.

# Conclusion

The RFE algorithm is a valuable feature selection method, especially for high-dimensional datasets with only a few observations.
The numerous settings of the algorithm in mlr3 make it possible to apply it to many datasets and learners.

> If you want to know more about feature selection in general, we recommend having a look at our [book](https://mlr3book.mlr-org.com/feature-selection.html.

Reviewer: ```suggestion
If you want to know more about feature selection in general, we recommend having a look at our [book](https://mlr3book.mlr-org.com/feature-selection.html).
```
